File: Semester 7/Computational Finance (MA 473)/Term Paper - Python/CCDScheme.py
import numpy as np
from compositeBoole import compositeBool
from findUInitial import findUInitial
from findVWInitial import findVWInitial
from g import g


def findCCDScheme(u_n, v_n, w_n, N, M, b, T, r, lambda_val, sigma, nu, eta):
    h = 2*b/M
    k = T/N
    x = np.arange(-b, b+h, h)

    B = np.zeros(3*M+3)
    A = np.zeros((3*M+3, 3*M+3))

    for i in range(0, 3*M-3):
        i1 = int((i+3)/3)
        if i % 3 == 2:
            gamma = compositeBool(u_n, M, b, r, lambda_val, sigma, nu, eta, i1)
            B[i] = u_n[i1] + k*sigma*sigma/4*w_n[i1] + k*lambda_val*gamma/2
        elif i % 3 == 0:
            B[i] = -v_n[i1]
        else:
            B[i] = 0

    gamma = compositeBool(u_n, M, b, r, lambda_val, sigma, nu, eta, 0)
    B[3*M-3] = u_n[0] + k*sigma*sigma/4*w_n[0] + k*lambda_val*gamma/2
    gamma = compositeBool(u_n, M, b, r, lambda_val, sigma, nu, eta, M)
    B[3*M-2] = u_n[M] + k*sigma*sigma/4*w_n[M] + k*lambda_val*gamma/2

    for i in range(0, 3*M-3):
        i1 = int((i+3)/3)
        if i % 3 == 2:
            A[i][i1] = 1
            A[i][2*M+2+i1] = -k*sigma*sigma/4
            A[i][0] += (2*h/45) * (-k*lambda_val/2) * 7 * \
                g(-b-x[i1], r, lambda_val, sigma, nu, eta)
            A[i][M] += (2*h/45) * (-k*lambda_val/2) * 7 * \
                g(b-x[i1], r, lambda_val, sigma, nu, eta)

            for l in range(0, int(M/4)-1):
                A[i][4*l+1] += (2*h/45) * (-k*lambda_val/2) * 32 * \
                    g(x[4*l+1]-x[i1], r, lambda_val, sigma, nu, eta)
                A[i][4*l+2] += (2*h/45) * (-k*lambda_val/2) * 12 * \
                    g(x[4*l+2]-x[i1], r, lambda_val, sigma, nu, eta)
                A[i][4*l+3] += (2*h/45) * (-k*lambda_val/2) * 32 * \
                    g(x[4*l+3]-x[i1], r, lambda_val, sigma, nu, eta)
                A[i][4*l+4] += (2*h/45) * (-k*lambda_val/2) * 14 * \
                    g(x[4*l+4]-x[i1], r, lambda_val, sigma, nu, eta)

            A[i][M-3] += (2*h/45) * (-k*lambda_val/2) * 32 * \
                g(x[M-3]-x[i1], r, lambda_val, sigma, nu, eta)
            A[i][M-2] += (2*h/45) * (-k*lambda_val/2) * 12 * \
                g(x[M-2]-x[i1], r, lambda_val, sigma, nu, eta)
            A[i][M-1] += (2*h/45) * (-k*lambda_val/2) * 32 * \
                g(x[M-1]-x[i1], r, lambda_val, sigma, nu, eta)

        elif i % 3 == 0:
            A[i][i1-1] = (15/(16*h))
            A[i][i1+1] = -(15/(16*h))
            A[i][M+1+i1-1] = 7/16
            A[i][M+1+i1+1] = 7/16
            A[i][2*M+2+i1-1] = h/16
            A[i][2*M+2+i1+1] = -h/16

        else:
            A[i][i1-1] = 3/(h*h)
            A[i][i1] = -6/(h*h)
            A[i][i1+1] = 3/(h*h)
            A[i][M+1+i1-1] = 9/(8*h)
            A[i][M+1+i1+1] = -9/(8*h)
            A[i][2*M+2+i1-1] = 1/8
            A[i][2*M+2+i1] = -1
            A[i][2*M+2+i1+1] = 1/8

    A[3*M-3][0] = 1
    A[3*M-3][2*M+2] = -k*sigma*sigma/4
    A[3*M-3][0] += (2*h/45) * (-k*lambda_val/2) * 7 * \
        g(-b-x[0], r, lambda_val, sigma, nu, eta)
    A[3*M-3][M] += (2*h/45) * (-k*lambda_val/2) * 7 * \
        g(b-x[0], r, lambda_val, sigma, nu, eta)

    for l in range(0, int(M/4)-1):
        A[3*M-3][4*l+1] += (2*h/45) * (-k*lambda_val/2) * \
            32 * g(x[4*l+1]-x[0], r, lambda_val, sigma, nu, eta)
        A[3*M-3][4*l+2] += (2*h/45) * (-k*lambda_val/2) * \
            12 * g(x[4*l+2]-x[0], r, lambda_val, sigma, nu, eta)
        A[3*M-3][4*l+3] += (2*h/45) * (-k*lambda_val/2) * \
            32 * g(x[4*l+3]-x[0], r, lambda_val, sigma, nu, eta)
        A[3*M-3][4*l+4] += (2*h/45) * (-k*lambda_val/2) * \
            14 * g(x[4*l+4]-x[0], r, lambda_val, sigma, nu, eta)

    A[3*M-3][M-3] += (2*h/45) * (-k*lambda_val/2) * 32 * \
        g(x[M-3]-x[0], r, lambda_val, sigma, nu, eta)
    A[3*M-3][M-2] += (2*h/45) * (-k*lambda_val/2) * 12 * \
        g(x[M-2]-x[0], r, lambda_val, sigma, nu, eta)
    A[3*M-3][M-1] += (2*h/45) * (-k*lambda_val/2) * 32 * \
        g(x[M-1]-x[0], r, lambda_val, sigma, nu, eta)

    # FOR j=M

    A[3*M-2][M] = 1
    A[3*M-2][2*M+2+M] = -k*sigma*sigma/4
    A[3*M-2][0] += (2*h/45) * (-k*lambda_val/2) * 7 * \
        g(-b-x[M], r, lambda_val, sigma, nu, eta)
    A[3*M-2][M] += (2*h/45) * (-k*lambda_val/2) * 7 * \
        g(b-x[M], r, lambda_val, sigma, nu, eta)

    for l in range(0, int(M/4)-1):
        A[3*M-2][4*l+1] += (2*h/45) * (-k*lambda_val/2) * \
            32 * g(x[4*l+1]-x[M], r, lambda_val, sigma, nu, eta)
        A[3*M-2][4*l+2] += (2*h/45) * (-k*lambda_val/2) * \
            12 * g(x[4*l+2]-x[M], r, lambda_val, sigma, nu, eta)
        A[3*M-2][4*l+3] += (2*h/45) * (-k*lambda_val/2) * \
            32 * g(x[4*l+3]-x[M], r, lambda_val, sigma, nu, eta)
        A[3*M-2][4*l+4] += (2*h/45) * (-k*lambda_val/2) * \
            14 * g(x[4*l+4]-x[M], r, lambda_val, sigma, nu, eta)

    A[3*M-2][M-3] += (2*h/45) * (-k*lambda_val/2) * 32 * \
        g(x[M-3]-x[M], r, lambda_val, sigma, nu, eta)
    A[3*M-2][M-2] += (2*h/45) * (-k*lambda_val/2) * 12 * \
        g(x[M-2]-x[M], r, lambda_val, sigma, nu, eta)
    A[3*M-2][M-1] += (2*h/45) * (-k*lambda_val/2) * 32 * \
        g(x[M-1]-x[M], r, lambda_val, sigma, nu, eta)

    A[3*M-1][0] = 31/h
    A[3*M-1][1] = -32/h
    A[3*M-1][2] = 1/h
    A[3*M-1][M+1] = 14
    A[3*M-1][M+1+1] = 16
    A[3*M-1][2*M+2] = 2*h
    A[3*M-1][2*M+2+1] = -4*h

    A[3*M][M] = -31/h
    A[3*M][M-1] = 32/h
    A[3*M][M-2] = -1/h
    A[3*M][M+1+M] = 14
    A[3*M][M+1+M-1] = 16
    A[3*M][2*M+2+M] = -2*h
    A[3*M][2*M+2+M-1] = 4*h

    A[3*M+1][0] = 7/(2*h)
    A[3*M+1][1] = -8/(2*h)
    A[3*M+1][2] = 1/(2*h)
    A[3*M+1][M+1] = 1
    A[3*M+1][M+1+1] = 2
    A[3*M+1][2*M+2+1] = -h

    A[3*M+2][M] = -7/(2*h)
    A[3*M+2][M-1] = 8/(2*h)
    A[3*M+2][M-2] = -1/(2*h)
    A[3*M+2][M+1+M] = 1
    A[3*M+2][M+1+M-1] = 2
    A[3*M+2][2*M+2+M-1] = h

    X = np.matmul(np.linalg.inv(A), B)

    u_n1 = X[0:M+1]
    v_n1 = X[M+1:2*M+2]
    w_n1 = X[2*M+2:]

    print(np.linalg.norm(np.matmul(A, X) - B))

    return u_n1, v_n1, w_n1


u_n = findUInitial(1.5, 3/128, 100, -0.9, 0.45, 0.05, 0.15, 0.1, 'call')
v_n, w_n = findVWInitial(u_n, 3/128, 128)

for i in range(0, 25):
    u_n, v_n, w_n = findCCDScheme(
        u_n, v_n, w_n, 25, 128, 1.5, 0.25, 0.05, 0.1, 0.15, -0.9, 0.45)


# print(u_n[60])

File: Semester 7/Computational Finance (MA 473)/MA473 Term Paper/Code/CCDScheme.m
function [u_n1, v_n1, w_n1] = CCDScheme(u_n, v_n, w_n, N, M, b, T, r, lambda, sigma, nu, eta, k_n)
%This function returns value of u, u' and u'' for
%time step n+1, given the values at time step n.
%Input:
%   u_n     : Array of values of u at time step n
%   v_n     : Array of values of u' at time step n
%   w_n     : Array of values of u'' at time step n
%   N       : Number of temporal steps
%   M       : Number of spatial steps
%   b       : |x| <= b
%   T       : Time to maturity
%   r       : Interest Rate
%   lambda  : Poisson rate of arrival of jumps
%   sigma   : std dev of Stock prices
%   nu      : mean of jump sizes
%   eta     : std dev of jump sizes
%   k_n     : iteration count       
%Output:
%   u_n1     : Array of values of u at time step n+1
%   v_n1     : Array of values of u' at time step n+1
%   w_n1     : Array of values of u'' at time step n+1
    
    %Calculating spatial and temporal step size and x
    h = (2*b)/M;
    k = T/N;
    x = -b:h:b;
    
    %Ax=B
    B=zeros(3*M+3,1);
    A=zeros(3*M+3,3*M+3);
    
    for i = 1:3*M-3
        i1 = floor((i+2)/3);
        if mod(i,3) == 0
            gamma = compositeBoole(u_n, M, b, r, lambda, sigma, nu, eta, i1+1);
            B(i) = u_n(i1+1)+(k*sigma*sigma/4)*w_n(i1+1)+(k*lambda/2)*gamma;
        elseif mod(i,3) == 1
            B(i) = -v_n(i1+1);
        else
            B(i) = 0;
        end
    end
    gamma = compositeBoole(u_n, M, b, r, lambda, sigma, nu, eta, 1);
    B(3*M-2) = u_n(1)+(k*sigma*sigma/4)*w_n(1)+(k*lambda/2)*gamma;
    gamma = compositeBoole(u_n, M, b, r, lambda, sigma, nu, eta, M+1);
    B(3*M-1) = u_n(M+1)+(k*sigma*sigma/4)*w_n(M+1)+(k*lambda/2)*gamma;
    %Last four values of B are zero because boundary conditions involve no
    %terms corresponding to n-th time step

    for i = 1:3*M-3
        i1 = floor((i+2)/3);
        if mod(i,3) == 0
            A(i, i1+1) = 1;
            A(i, 2*M+2+i1+1) = -k*sigma*sigma/4;
            A(i, 1) = A(i, 1) + (-k*h*lambda/45)*(7*g(-b-x(i1+1), r, lambda, sigma, nu, eta));
            A(i, M+1) = A(i, M+1) + (-k*h*lambda/45)*(7*g(b-x(i1+1), r, lambda, sigma, nu, eta));
          
            for j = 1 : M/4-1
                A(i, 4*j-2) = A(i, 4*j-2) + (-k*h*lambda/45)*32*g(x(4*j-2)-x(i1+1), r, lambda, sigma, nu, eta);
                A(i, 4*j-1) = A(i, 4*j-1) + (-k*h*lambda/45)*12*g(x(4*j-1)-x(i1+1), r, lambda, sigma, nu, eta);
                A(i, 4*j) = A(i, 4*j) + (-k*h*lambda/45)*32*g(x(4*j)-x(i1+1), r, lambda, sigma, nu, eta);
                A(i, 4*j+1) = A(i, 4*j+1) + (-k*h*lambda/45)*14*g(x(4*j+1)-x(i1+1), r, lambda, sigma, nu, eta);
            end
            
            A(i, M-2) = A(i, M-2) + (-k*h*lambda/45)*32*g(x(M-2)-x(i1+1), r, lambda, sigma, nu, eta);
            A(i, M-1) = A(i, M-1) + (-k*h*lambda/45)*12*g(x(M-1)-x(i1+1), r, lambda, sigma, nu, eta);
            A(i, M) = A(i, M) + (-k*h*lambda/45)*32*g(x(M)-x(i1+1), r, lambda, sigma, nu, eta);

        elseif mod(i,3) == 1
            A(i, i1) = 15/(16*h);
            A(i, i1+2) = -15/(16*h);
            A(i, M+1+i1) = 7/16;
            A(i, M+1+i1+2) = 7/16;
            A(i, 2*M+2+i1) = h/16;
            A(i, 2*M+2+i1+2) = -h/16;
        else
            A(i, i1) = 3/(h*h);
            A(i, i1+1) = -6/(h*h);
            A(i, i1+2) = 3/(h*h);
            A(i, M+1+i1) = 9/(8*h);
            A(i, M+1+i1+2) = -9/(8*h);
            A(i, 2*M+2+i1) = 1/8;
            A(i, 2*M+2+i1+1) = -1;
            A(i, 2*M+2+i1+2) = 1/8;
        end
    end

    %Equation 1 for j = 0
    A(3*M-2, 2*M+2+1) = -k*sigma*sigma/4;
    
    A(3*M-2, 1) = 1 + (-k*h*lambda/45)*(7*g(0, r, lambda, sigma, nu, eta));
    A(3*M-2, M+1) = A(3*M-2, M+1) + (-k*h*lambda/45)*(7*g(2*b, r, lambda, sigma, nu, eta));
  
    for j = 1 : M/4-1
        A(3*M-2, 4*j-2) = A(3*M-2, 4*j-2) + (-k*h*lambda/45)*32*g(x(4*j-2)+b, r, lambda, sigma, nu, eta);
        A(3*M-2, 4*j-1) = A(3*M-2, 4*j-1) + (-k*h*lambda/45)*12*g(x(4*j-1)+b, r, lambda, sigma, nu, eta);
        A(3*M-2, 4*j) = A(3*M-2, 4*j) + (-k*h*lambda/45)*32*g(x(4*j)+b, r, lambda, sigma, nu, eta);
        A(3*M-2, 4*j+1) = A(3*M-2, 4*j+1) + (-k*h*lambda/45)*14*g(x(4*j+1)+b, r, lambda, sigma, nu, eta);
    end
    
    A(3*M-2, M-2) = A(3*M-2, M-2) + (-k*h*lambda/45)*32*g(x(M-2)+b, r, lambda, sigma, nu, eta);
    A(3*M-2, M-1) = A(3*M-2, M-1) + (-k*h*lambda/45)*12*g(x(M-1)+b, r, lambda, sigma, nu, eta);
    A(3*M-2, M) = A(3*M-2, M) + (-k*h*lambda/45)*32*g(x(M)+b, r, lambda, sigma, nu, eta);

    %Equation 1 for j = M
    A(3*M-1, 3*M+3) = -k*sigma*sigma/4;
    A(3*M-1, 1) = A(3*M-1, 1) + (-k*h*lambda/45)*(7*g(-2*b, r, lambda, sigma, nu, eta));
    A(3*M-1, M+1) = 1 + (-k*h*lambda/45)*(7*g(0, r, lambda, sigma, nu, eta));
    for j = 1 : M/4-1
        A(3*M-1, 4*j-2) = A(3*M-1, 4*j-2) + (-k*h*lambda/45)*32*g(x(4*j-2)-b, r, lambda, sigma, nu, eta); 
        A(3*M-1, 4*j-1) = A(3*M-1, 4*j-1) + (-k*h*lambda/45)*12*g(x(4*j-1)-b, r, lambda, sigma, nu, eta);
        A(3*M-1, 4*j) = A(3*M-1, 4*j) + (-k*h*lambda/45)*32*g(x(4*j)-b, r, lambda, sigma, nu, eta);
        A(3*M-1, 4*j+1) = A(3*M-1, 4*j+1) + (-k*h*lambda/45)*14*g(x(4*j+1)-b, r, lambda, sigma, nu, eta);
    end
    
    A(3*M-1, M-2) = A(3*M-1, M-2) + (-k*h*lambda/45)*32*g(x(M-2)-b, r, lambda, sigma, nu, eta);
    A(3*M-1, M-1) = A(3*M-1, M-1) + (-k*h*lambda/45)*12*g(x(M-1)-b, r, lambda, sigma, nu, eta);
    A(3*M-1, M) = A(3*M-1, M) + (-k*h*lambda/45)*32*g(x(M)-b, r, lambda, sigma, nu, eta);

    %Equation 4
    A(3*M, 1) = 31/h; A(3*M, 2) = -32/h; A(3*M, 3) = 1/h; 
    A(3*M, M+1+1) = 14; A(3*M, M+1+2) = 16; 
    A(3*M, 2*M+2+1) = 2*h; A(3*M, 2*M+2+2) = -4*h;

    %Equation 5
    A(3*M+1, M+1) = -31/h; A(3*M+1, M) = 32/h; A(3*M+1, M-1) = -1/h; 
    A(3*M+1, M+1+M+1) = 14;A(3*M+1, M+1+M) = 16; 
    A(3*M+1, 2*M+2+M+1) = -2*h; A(3*M+1, 2*M+2+M) = 4*h;

    %Equation 6
    A(3*M+2, 1) = 7/(2*h); A(3*M+2, 2) = -4/h; A(3*M+2, 3) = 1/(2*h);
    A(3*M+2, M+1+1) = 1; A(3*M+2, M+1+2) = 2; 
    A(3*M+2, 2*M+2+2) = -h;

    %Equation 7
    A(3*M+3, M+1) = -7/(2*h); A(3*M+3, M) = 4/h; A(3*M+3, M-1) = -1/(2*h);
    A(3*M+3, M+1+M+1) = 1; A(3*M+3, M+1+M) = 2;
    A(3*M+3, 2*M+2+M) = h;
    
    u0 = approxSol(x, N, T, r, lambda, sigma, nu, eta, k_n);
    [v0, w0] = findVWInitial(u0,h,M);
    x0 = [u0;v0;w0];

    [sol, ~] = bicgstabl(A,B,1e-6,1000,[],[],x0);

    u_n1 = sol(1:M+1);
    v_n1 = sol(M+2:2*M+2);
    w_n1 = sol(2*M+3:end);
end